pushsync/simulation_test.go

```diff
@@ -0,0 +1,292 @@
+	return err
+}
+
+// newServiceFunc constructs a minimal service needed for a simulation test for Push Sync, namely:
+// localstore, netstore, stream and pss
+func newServiceFunc(ctx *adapters.ServiceContext, bucket *sync.Map) (node.Service, func(), error) {
+	// setup localstore
+	n := ctx.Config.Node()
+	addr := network.NewAddr(n)
+	dir, err := ioutil.TempDir("", "localstore-test")
+	if err != nil {
+		return nil, nil, err
+	}
+	lstore, err := localstore.New(dir, addr.Over(), nil)
+	if err != nil {
+		os.RemoveAll(dir)
+		return nil, nil, err
+	}
+	// setup netstore
+	netStore := storage.NewNetStore(lstore, enode.ID{})
+
+	// setup pss
+	kadParams := network.NewKadParams()
+	kad := network.NewKademlia(addr.Over(), kadParams)
+	bucket.Store(simulation.BucketKeyKademlia, kad)
+
+	privKey, err := crypto.GenerateKey()
+	pssp := pss.NewParams().WithPrivateKey(privKey)
+	ps, err := pss.New(kad, pssp)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// // streamer for retrieval
+	delivery := stream.NewDelivery(kad, netStore)
+	netStore.RemoteGet = delivery.RequestFromPeers
+
+	bucket.Store(bucketKeyNetStore, netStore)
+
+	// set up syncer
+	noSyncing := &stream.RegistryOptions{Syncing: stream.SyncingDisabled, SyncUpdateDelay: 50 * time.Millisecond}
+	r := stream.NewRegistry(addr.ID(), delivery, netStore, state.NewInmemoryStore(), noSyncing, nil)
+
+	pubSub := pss.NewPubSub(ps)
+	// setup pusher
+	p := NewPusher(lstore, pubSub, chunk.NewTags())
+	bucket.Store(bucketKeyPushSyncer, p)
+
+	// setup storer
+	s := NewStorer(netStore, pubSub)
+
+	cleanup := func() {
+		r.Close()
+		p.Close()
+		s.Close()
+		netStore.Close()
+		os.RemoveAll(dir)
+	}
+
+	return &StreamerAndPss{r, ps}, cleanup, nil
+}
+
+// implements the node.Service interface
+type StreamerAndPss struct {
+	*stream.Registry
+	pss *pss.Pss
+}
+
+func (s *StreamerAndPss) Protocols() []p2p.Protocol {
+	return append(s.Registry.Protocols(), s.pss.Protocols()...)
+}
+
+func (s *StreamerAndPss) Start(srv *p2p.Server) error {
+	err := s.Registry.Start(srv)
+	if err != nil {
+		return err
+	}
+	return s.pss.Start(srv)
+}
+
+func (s *StreamerAndPss) Stop() error {
+	err := s.Registry.Stop()
+	if err != nil {
+		return err
+	}
+	return s.pss.Stop()
+}
+
+func upload(ctx context.Context, store Store, tags *chunk.Tags, tagname string, n int) (tag *chunk.Tag, addrs []storage.Address, err error) {
+	tag, err = tags.Create(tagname, int64(n))
+	if err != nil {
+		return nil, nil, err
+	}
+	for i := 0; i < n; i++ {
+		ch := storage.GenerateRandomChunk(int64(chunk.DefaultSize))
+		addrs = append(addrs, ch.Address())
+		_, err := store.Put(ctx, chunk.ModePutUpload, ch.WithTagID(tag.Uid))
+		if err != nil {
+			return nil, nil, err
+		}
+		tag.Inc(chunk.StateStored)
+	}
+	return tag, addrs, nil
+}
+
+func download(ctx context.Context, store *storage.NetStore, addrs []storage.Address) error {
+	errc := make(chan error)
+	for _, addr := range addrs {
```

Comment:
please add a `WaitGroup` here

Reply:
I think that `i` and break in loop below is handling the correct waiting.

Reply:
this can actually be written with `errGroup`:
```
var g errgroup.Group
for _, addr := range addrs {
    addr := addr
    g.Go(func() error {
	_, err := store.Get(ctx, chunk.ModeGetRequest, storage.NewRequest(addr))
	log.Debug("Get", "addr", hex.EncodeToString(addr[:]), "err", err)
        return err
    })
}
return g.Wait()
```
more [here](https://godoc.org/golang.org/x/sync/errgroup)